lib/ui/widgets/ata_map.dart

```diff
@@ -0,0 +1,186 @@
+import 'dart:async';
+import 'package:flutter/material.dart';
+import 'package:geolocator/geolocator.dart';
+import 'package:google_maps_flutter/google_maps_flutter.dart';
+import 'package:flutter/services.dart';
+import 'package:maps_toolkit/maps_toolkit.dart' as maps;
+
+/// A custom Flutter Google Map.
+/// Have some features:
+/// 1. Specify current location of device
+/// 2. Pin any marker on map and move it
+/// 3. Show circle around marked position with deviation radius
+/// 4. Calulate distance between current location and marked position
+/// Example code :
+/// ==> admin user
+/// ```dart
+/// ATAMap(
+///    isMoveableMarker: true,
+///)
+/// ```
+/// ==> normal user
+/// ```dart
+/// ATAMap(
+///    titleMarkedPosition: 'Office Position',
+///    markedLat: 10.762622,
+///    markedLng:  106.660172,
+///)
+/// ```
+class ATAMap extends StatefulWidget {
+  final double centerMapLat;
+  final double centerMapLng;
+  final double markedLat;
+  final double markedLng;
+  final bool isMoveableMarker;
+  final double deviationRadius;
+  final String titleMarkedPosition;
+  ATAMap(
+      {this.markedLat = 10.7440878,
+      this.markedLng = 106.7007886,
+      this.centerMapLat = 10.7440878,
+      this.centerMapLng = 106.7007886,
+      this.isMoveableMarker = false,
+      this.deviationRadius = 100,
+      this.titleMarkedPosition = "Marked Position"});
+
+  @override
+  State<ATAMap> createState() => ATAMapState(
+      currentMarkedLat: markedLat,
+      currentMarkedLng: markedLng,
+      centerMapLat: centerMapLat,
+      centerMapLng: centerMapLng,
+      isMoveableMarker: isMoveableMarker,
+      deviationRadius: deviationRadius,
+      titleMarkedPosition: titleMarkedPosition);
+}
+
+class ATAMapState extends State<ATAMap> {
+  CameraPosition defaultCamera;
+  static const double DEFAULT_ZOOM = 17;
+  //static const double deviationRadius = 100;
+  static const int TIMEOUT_PIN_MARKER_MAP_LOADED = 4;
+  Completer<GoogleMapController> _controller = Completer();
+
+  Set<Marker> _markers = Set();
+  Set<Circle> _circles = Set();
+  final double centerMapLat;
+  final double centerMapLng;
+  double currentLocationLat;
+  double currentLocationLng;
+  double currentMarkedLat;
+  double currentMarkedLng;
+  double deviationRadius;
+
+  BitmapDescriptor markedLocationIcon;
+  final bool isMoveableMarker;
+  final String titleMarkedPosition;
+
+  ATAMapState({this.currentMarkedLat, this.currentMarkedLng, this.centerMapLat, this.centerMapLng, this.isMoveableMarker, this.deviationRadius, this.titleMarkedPosition}) {

```

Comment:
no need to add params to State constructor just to pass values from Widget to its state.

inside the State<> class, we can always access the widget's props through widget.markedLat